### PR TITLE
Public offer strict encoding

### PR DESCRIPTION
**Pull request by Lederstrumpf**

### CODE REVIEW

**Comment by h4sh3d** [comment]
I'm not convinced the best way for public offer serde is with the string, if we want a string we use a String type and have it with `offer.to_string()` when we want to serialize the encoded offer. I'd prefer having proper serde impl for all the fields, this is something I've worked on but not finished.

**Comment by Lederstrumpf** [comment]
Fully agree it's not the best way. I suggest we use this and switch to your proper serde impl once that's done, unless you can finish that quickly?

File: src/hash.rs [hunk 1]
```diff
@@ -31,3 +31,30 @@ impl<'de> Visitor<'de> for HashString {
...
+            formatter,
+            "a string representing a public offer in Base58 value"
+        )
+    }
+
+    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        if s.len() == 195 {
```
Comment (h4sh3d) [inline src/hash.rs:54]: Are we sure that the len is always 195, my take is not as base58 can vary depending on values

File: src/negotiation.rs [hunk 2]
```diff
@@ -583,6 +583,39 @@ where
...
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s: String = PUB_OFFER_PREFIX.to_owned();
+        let encoded = base58_monero::encode_check(consensus::serialize(self).as_ref())
+            .expect("Encoding in base58 check works");
+        s.push_str(&encoded);
+        serializer.serialize_str(s.as_ref())
```
Comment (h4sh3d) [inline src/negotiation.rs:599]: This code will also run if we use `to_string`, we should use it to maintain only one interface IMO

Comment (Lederstrumpf) [reply src/negotiation.rs]: agreed - updated

File: src/hash.rs (hunk 1 above)
Comment (Lederstrumpf) [reply src/hash.rs:54]: No, not sure about this - are there any variable size members of `PublicOffer`?

Comment (h4sh3d) [reply src/hash.rs:54]: The way it currently works yes, depending on the type of assets you may have different bytes serialized (but with only xmr/btc we are fine so we can keep it).

**Review by h4sh3d — APPROVED** [review]

**Pull request merged**